### PR TITLE
chore: push protocol error conversion to dispatcher

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: packages/playwright-core/src/server/chromium/crConnection.ts [hunk 1]
```diff
@@ -199,10 +189,10 @@ export class CRSession extends EventEmitter {
   dispose() {
     this._closed = true;
     this._connection._sessions.delete(this._sessionId);
-    const errorMessage = this._closedErrorMessage()!;
     for (const callback of this._callbacks.values()) {
-      callback.error.sessionClosed = true;
-      callback.reject(rewriteErrorMessage(callback.error, errorMessage));
+      callback.error.type = 'closed';
```
Comment (dgozman) [inline packages/playwright-core/src/server/chromium/crConnection.ts:193]:
```suggestion
      callback.error.type = this._crashed ? 'crashed' : 'closed';
```

File: packages/playwright-core/src/server/protocolError.ts [hunk 2]
```diff
@@ -14,15 +14,33 @@
...
 
-  constructor(sessionClosed: boolean, message?: string) {
-    super(message);
-    this.sessionClosed = sessionClosed || false;
+  setMessage(message: string) {
+    rewriteErrorMessage(this, `Protocol error (${this.method}): ${message}`);
   }
+
+  browserLogMessage() {
+    return this.logs ? '\nBrowser logs:\n' + this.logs : '';
```
Comment (dgozman) [inline packages/playwright-core/src/server/protocolError.ts:36]: Given that we use this for xserver and stuff, perhaps rename it to just `logs` and move the `'Browser logs'` prefix to the constructor call site?

File: packages/playwright-core/src/server/webkit/wkConnection.ts [hunk 3]
```diff
@@ -159,11 +154,10 @@ export class WKSession extends EventEmitter {
   }
 
   dispose() {
-    if (this.connection._browserDisconnectedLogs)
-      this.errorText = kTargetClosedErrorMessage + '\nBrowser logs: ' + this.connection._browserDisconnectedLogs;
     for (const callback of this._callbacks.values()) {
-      callback.error.sessionClosed = true;
-      callback.reject(rewriteErrorMessage(callback.error, this.errorText));
+      callback.error.type = 'closed';
```
Comment (dgozman) [inline packages/playwright-core/src/server/webkit/wkConnection.ts:158]: ditto?

File: packages/playwright-core/src/server/webkit/wkConnection.ts [hunk 4]
```diff
@@ -121,7 +119,6 @@ export class WKSession extends EventEmitter {
     this.connection = connection;
     this.sessionId = sessionId;
     this._rawSend = rawSend;
-    this.errorText = errorText;
```
Comment (dgozman) [inline packages/playwright-core/src/server/webkit/wkConnection.ts:124]: Drop the `errorText` parameter then?

File: packages/playwright-core/src/server/browserType.ts [hunk 5]
```diff
@@ -294,10 +295,16 @@ export abstract class BrowserType extends SdkObject {
     }
   }
 
+  _rewriteStartupError(error: Error): Error {
+    if (!isProtocolError(error))
+      return error;
+    return this._doRewriteStartupError(error);
```
Comment (dgozman) [inline packages/playwright-core/src/server/browserType.ts:301]: Are you sure it is always a protocol error? Could it be something else, when process suddenly exits for example?

File: packages/playwright-core/src/server/firefox/ffConnection.ts [hunk 6]
```diff
@@ -178,18 +168,11 @@ export class FFSession extends EventEmitter {
   dispose() {
     this._disposed = true;
     this._connection._sessions.delete(this._sessionId);
-    const errorMessage = this._closedErrorMessage()!;
     for (const callback of this._callbacks.values()) {
-      callback.error.sessionClosed = true;
-      callback.reject(rewriteErrorMessage(callback.error, errorMessage));
+      callback.error.type = 'closed';
```
Comment (dgozman) [inline packages/playwright-core/src/server/firefox/ffConnection.ts:172]: ditto?

**Review by dgozman — APPROVED** [review]

File: packages/playwright-core/src/server/protocolError.ts (hunk 2 above)
Comment (pavelfeldman) [reply packages/playwright-core/src/server/protocolError.ts:36]: Let me follow up

**Pull request merged**